engineio/src/client/socket.rs

```diff
@@ -0,0 +1,606 @@
+use super::super::socket::Socket as InnerSocket;
+use crate::callback::OptionalCallback;
+use crate::transport::Transport;
+
+use crate::error::{Error, Result};
+use crate::header::HeaderMap;
+use crate::packet::{HandshakePacket, Packet, PacketId};
+use crate::transports::{PollingTransport, WebsocketSecureTransport, WebsocketTransport};
+use bytes::Bytes;
+use native_tls::TlsConnector;
+use std::convert::TryFrom;
+use std::convert::TryInto;
+use std::fmt::Debug;
+use url::Url;
+
+#[derive(Clone, Debug)]
+pub struct Socket {
+    socket: InnerSocket,
+}
+
+#[derive(Clone, Debug)]
+pub struct SocketBuilder {
+    url: Url,
+    tls_config: Option<TlsConnector>,
+    headers: Option<HeaderMap>,
+    handshake: Option<HandshakePacket>,
+    on_error: OptionalCallback<String>,
+    on_open: OptionalCallback<()>,
+    on_close: OptionalCallback<()>,
+    on_data: OptionalCallback<Bytes>,
+    on_packet: OptionalCallback<Packet>,
+}
+
+impl SocketBuilder {
+    pub fn new(url: Url) -> Self {
+        let mut url = url;
+        url.query_pairs_mut().append_pair("EIO", "4");
```

Comment:
Should we move this into a static variable? If we might change the version some time globally?

Reply:
https://github.com/1c3t3a/rust-socketio/pull/117